src/main/java/org/opensearch/knn/index/codec/transfer/OffHeapQuantizedVectorTransfer.java

```diff
@@ -0,0 +1,149 @@
+ */
+
+package org.opensearch.knn.index.codec.transfer;
+
+import lombok.Getter;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.opensearch.knn.common.KNNVectorUtil.createArrayList;
+
+/**
+ * The class is intended to transfer {@link KNNVectorValues} to off heap memory. If also provides and ability to quantize the vector
+ * before it is transferred to offHeap memory. The class is not thread safe
+ *
+ * @param <T> an array of primitive type
+ * @param <V> an array of primitive type after being quantized
+ */
+abstract class OffHeapQuantizedVectorTransfer<T, V> implements VectorTransfer {
+
+    protected static final int DEFAULT_COMPRESSION_FACTOR = 1;
+
+    @Getter
+    private long vectorAddress;
+    @Getter
+    private int[] transferredDocsIds;
+    private final int transferLimit;
+    // Keeping this as a member variable as this should not be changed considering the vector address is reused between batches
+    protected long batchSize;
+    protected final long bytesPerVector;
+
+    private final List<V> vectorsToTransfer;
+    private final List<Integer> transferredDocIdsList;
+
+    private final KNNVectorValues<T> vectorValues;
+
+    // TODO: Replace with actual quantization parameters
+    private final BiFunction<T, String, V> quantizer;
+    private final String quantizationState;
+
+    private final int compressionFactor;
+
+    public OffHeapQuantizedVectorTransfer(
+        final KNNVectorValues<T> vectorValues,
+        final Long batchSize,
+        final BiFunction<T, String, V> quantizer,
+        final String quantizationState,
+        final int compressionFactor
+    ) {
+        assert vectorValues.docId() != -1 : "vectorValues docId must be set, iterate it once for vector transfer to succeed";
+        assert vectorValues.docId() != NO_MORE_DOCS : "vectorValues already iterated, Nothing to transfer";
+
+        this.quantizer = quantizer;
+        this.quantizationState = quantizationState;
+        this.bytesPerVector = vectorValues.bytesPerVector() / compressionFactor;
+        this.compressionFactor = compressionFactor;
+        this.transferLimit = (int) Math.max(1, (int) KNNSettings.getVectorStreamingMemoryLimit().getBytes() / bytesPerVector);
+        this.batchSize = batchSize == null ? transferLimit : batchSize;
+        this.vectorsToTransfer = createArrayList(this.batchSize);
+        this.transferredDocIdsList = createArrayList(this.batchSize);
+        this.vectorValues = vectorValues;
+        this.vectorAddress = 0; // we can allocate initial memory here, currently storeVectorData takes care of it
+    }
+
+    @Override
+    public void transfer() throws IOException {
+        if (vectorValues.docId() == NO_MORE_DOCS) {
```

Comment:
why we are doing this check? Plus this transfer function is only called from Transferbatch do we need to make this as public? Seeing public I am trying to see who is calling this function

Reply:
> why we are doing this check?

Its being defensive so client does not go into infinite loop

>  Plus this transfer function is only called from Transferbatch do we need to make this as public? Seeing public I am trying to see who is calling this function

https://github.com/opensearch-project/k-NN/pull/1950#discussion_r1715970582

Reply:
Lets mention this in the java doc. Just to know that why these conditions are there.

Reply:
There is a comment there, anywhere else you want me to put it?